### PR TITLE
codec(ticdc): simple protocol support encode dml event

### LINKED ISSUE #9902: encoding & decoding DML event

### LINKED ISSUE #10185: add new protocol consumer

### LINKED ISSUE #10186: add integration test cases for new protocol

### PR DESCRIPTION
**Pull request by 3AceShowHand**
### What problem does this PR solve?

Issue Number: close #9902, close #10186, close #10185, 

### What is changed and how it works?

* simple protocol support encode and decode DML event.
* entry schema test helper support convert DDL to DDLEvent, DML to RowChangedEvent
* add simple basic integration test

#### Tests 

 - Unit test
 - Integration test

### Release note 

```release-note
`None`
```

### CODE REVIEW

File: pkg/sink/codec/simple/decoder.go [hunk 1]
```diff
@@ -78,8 +82,27 @@ func (d *decoder) NextResolvedEvent() (uint64, error) {
 
 // NextRowChangedEvent returns the next row changed event if exists
 func (d *decoder) NextRowChangedEvent() (*model.RowChangedEvent, error) {
-	// TODO implement me
-	panic("implement me")
+	if d.msg == nil || (d.msg.Data == nil && d.msg.Old == nil) {
+		return nil, cerror.ErrCodecDecode.GenWithStack(
+			"not found row changed event message")
```
Comment (asddongmen) [inline pkg/sink/codec/simple/decoder.go:87]:
```suggestion
			"invalid row changed event message")
```

File: pkg/sink/codec/simple/message.go [hunk 2]
```diff
@@ -48,165 +56,141 @@ const (
...
 	for _, col := range index.Columns {
-		i.Columns = append(i.Columns, col.Name.O)
-	}
-
-	for _, col := range index.Columns {
-		offset := col.Offset
-		colInfo := tableInfo.Columns[offset]
+		indexSchema.Columns = append(indexSchema.Columns, col.Name.O)
+		colInfo := columns[col.Offset]
 		if mysql.HasNotNullFlag(colInfo.GetFlag()) {
```
Comment (asddongmen) [inline pkg/sink/codec/simple/message.go:140]: This if block should be put in a separated for loop.

File: pkg/sink/codec/simple/message.go [hunk 3]
```diff
@@ -48,165 +56,141 @@ const (
...
-			i.Nullable = "false"
+			indexSchema.Nullable = false
 			break
```
Comment (asddongmen) [inline pkg/sink/codec/simple/message.go:142]: Can not break in this for loop.

File: pkg/sink/codec/simple/message.go [hunk 4]
```diff
@@ -48,165 +56,141 @@ const (
...
+	return indexSchema
 }
 
-// ToTiIndexInfo converts from Index to TiDB IndexInfo.
-func (i *IndexSchema) ToTiIndexInfo() (*timodel.IndexInfo, error) {
-	index := new(timodel.IndexInfo)
-	index.Columns = make([]*timodel.IndexColumn, 0)
-	for i, col := range i.Columns {
-		index.Columns = append(index.Columns, &timodel.IndexColumn{
+func newTiIndexInfo(indexSchema *IndexSchema) *timodel.IndexInfo {
```
Comment (asddongmen) [inline pkg/sink/codec/simple/message.go:148]:
```suggestion
func newTiDBIndexInfo(indexSchema *IndexSchema) *timodel.IndexInfo {
```

File: pkg/sink/codec/utils/mysql_types.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package utils
```
Comment (CharlesCheung96) [inline pkg/sink/codec/utils/mysql_types.go:14]: Please add tests for mysql_types.go

File: pkg/sink/codec/simple/decoder.go [hunk 6]
```diff
@@ -88,10 +116,67 @@ func (d *decoder) NextDDLEvent() (*model.DDLEvent, error) {
...
+
+func (m *memoryTableInfoProvider) Write(info *model.TableInfo) {
+	key := cacheKey{
+		schema: info.TableName.Schema,
+		table:  info.TableName.Table,
+	}
+
+	entry, ok := m.memo[key]
+	if ok && entry.UpdateTS >= info.UpdateTS {
+		log.Warn("table info not stored, since the updateTs is stale",
```
Comment (asddongmen) [inline pkg/sink/codec/simple/decoder.go:152]:
```suggestion
		log.Warn("refuse to store this tableInfo, as its updateTs is less than current one",
```

File: pkg/sink/codec/simple/decoder.go [hunk 7]
```diff
@@ -88,10 +116,67 @@ func (d *decoder) NextDDLEvent() (*model.DDLEvent, error) {
...
+	memo map[cacheKey]*model.TableInfo
+}
+
+func newMemoryTableInfoProvider() *memoryTableInfoProvider {
+	return &memoryTableInfoProvider{
+		memo: make(map[cacheKey]*model.TableInfo),
+	}
+}
+
+func (m *memoryTableInfoProvider) Write(info *model.TableInfo) {
```
Comment (asddongmen) [inline pkg/sink/codec/simple/decoder.go:144]: Add comment for exported method.

File: pkg/sink/codec/simple/message.go [hunk 8]
```diff
@@ -274,31 +302,168 @@ func newResolvedMessage(ts uint64) *message {
...
 func newDDLMessage(ddl *model.DDLEvent) *message {
-	var tableSchema TableSchema
-	tableSchema.FromTableInfo(ddl.TableInfo)
+	var (
+		database      string
+		table         string
+		schema        *TableSchema
+		schemaVersion uint64
+	)
+	// the tableInfo maybe nil if the DDL is `drop database`
```
Comment (asddongmen) [inline pkg/sink/codec/simple/message.go:311]:
```suggestion
	// the tableInfo maybe nil if the DDL is a schema level DDL, like `drop database`, `create database`
```

File: pkg/sink/codec/simple/message.go [hunk 9]
```diff
@@ -48,204 +56,222 @@ const (
...
-	Unique   string   `json:"unique,omitempty"`
-	Primary  string   `json:"primary,omitempty"`
-	Nullable string   `json:"nullable,omitempty"`
+	Unique   bool     `json:"unique"`
+	Primary  bool     `json:"primary"`
+	Nullable bool     `json:"nullable"`
 	Columns  []string `json:"columns"`
 }
 
 // FromTiIndexInfo converts from TiDB IndexInfo to Index.
```
Comment (asddongmen) [inline pkg/sink/codec/simple/message.go:129]:
```suggestion
// newIndexSchema converts from TiDB IndexInfo to IndexSchema.
```

**Review by asddongmen — APPROVED** [review]

File: pkg/sink/codec/utils/mysql_types.go (hunk 5 above)
Comment (3AceShowHand) [reply pkg/sink/codec/utils/mysql_types.go:14]: this is covered by other unit test, if this is incorrect,  canal-json unit test will be failed.

File: cdc/owner/ddl_manager.go [hunk 10]
```diff
@@ -208,7 +208,7 @@ func (m *ddlManager) tick(
 			return nil, nil, err
 		}
 		if !ok {
-			return nil, nil, nil
+			return nil, m.barrier(), nil
```
Comment (CharlesCheung96) [inline cdc/owner/ddl_manager.go:211]: Why return an empty barrier here?

File: cdc/owner/ddl_manager.go [hunk 11]
```diff
@@ -208,7 +208,7 @@ func (m *ddlManager) tick(
 			return nil, nil, err
 		}
 		if !ok {
```
Comment (CharlesCheung96) [inline cdc/owner/ddl_manager.go:210]:
```suggestion
		if !finished {
```

File: cdc/owner/ddl_manager.go (hunk 10 above)
Comment (3AceShowHand) [reply cdc/owner/ddl_manager.go:211]: return nil cause the panic, and the barrier is initialized as empty I think.

**Review by CharlesCheung96 — APPROVED** [review]

**Pull request merged**